KIPs/kip-xxx.md

```diff
@@ -0,0 +1,182 @@
+| `EMPTY_CODE_HASH` | `0xc5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470` |
+| `SET_CODE_TX_TYPE` | `0x04` |
+| `ENVELOPED_SET_CODE_TX_TYPE` | `0x7804` |
+
+### SetCode Transaction
+
+Introduce the SetCode transaction type where its RLP encoding is defined below.
+
+```
+0x7804 || rlp([chain_id, nonce, max_priority_fee_per_gas, max_fee_per_gas, gas_limit, destination, value, data, access_list, authorization_list, signature_y_parity, signature_r, signature_s])
+
+authorization_list = [[chain_id, address, nonce, y_parity, r, s], ...]
+```
+
+The fields follow the same semantics as EIP-7702.
+
+Its consensus type prefix (that affects `block.transactionRoot`) is `ENVELOPED_SET_CODE_TX_TYPE (0x7804)` and its eth namespace API type prefix is `SET_CODE_TX_TYPE (0x04)`. See “Changes to API” for clarification.
+
+#### Behavior
+
+The transaction’s behavior follows that of EIP-7702 except for the following:
+
+- For each authorization tuple `[chain_id, address, nonce, y_parity, r, s]`, fetch the AccontKey field of the recovered authority account.
+- Verify that the AccountKey is of type `AccountKeyTypeLegacy`. If verification fails, immediately stop processing the tuple and continue to the next tuple in the list.
+
+#### Delegation Designation
+
+Semantics of `EXTCODESIZE`, `EXTCODECOPY`, `EXTCODEHASH`, `CALL`, `CALLCODE`, `STATICCALL`, `DELEGATECALL` opcodes follows EIP-7702.
+
+#### Gas costs
+
+The intrinsic gas cost of a SetCode transaction follows the EIP-7702, meaning it inherits [EIP-2930](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-2930.md) and [EIP-2028](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-2028.md). Note that the EIP-2028 rule is realized by by KIP-xxx.
+
+#### Transaction Origination
+
+Since Kaia hasn't been explicitly enforcing [EIP-3607](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-3607.md), there is no need to modify transaction origin restriction.
+
+Do follow the EIP-7702’s behavior where if a transaction’s destination has a delegation, also warm its target
+
+### Account State
+
+The account RLP encoding for the world state trie needs to change.
+
+```
+EOA's compact encoding: 0x01 || rlp([nonce, balance, humanReadable, accountKey])
+EOA's full encoding:    0x01 || rlp([[nonce, balance, humanReadable, accountKey], storageRoot, codeHash, codeInfo])
+SCA's encoding:         0x02 || rlp([[nonce, balance, humanReadable, accountKey], storageRoot, codeHash, codeInfo])
+```
+
+Now both EOAs and SCAs hold the same set of 7 fields: `nonce`, `balance`, `humanReadable`, `accountKey`, `storageRoot`, `codeHash`, and `codeInfo`.
+
+When an EOA is RLP encoded, the compact encoding is used when the account has storageRoot = EMPTY_ROOT_HASH, codeHash = EMPTY_CODE_HASH, and codeInfo = 0x0. Conversely, when decoding an RLP-encoded EOA, the compact encoding implies that the account has storageRoot = EMPTY_ROOT_HASH, codeHash = EMPTY_CODE_HASH, and codeInfo = 0x0.
```

Comment:
If it performs an initialize that EOA does not expect, golang will result in storageRoot = 0x0, codeHash = nil, and codeInfo = 0x0.
Compact encoding will be applied in this case as well, but is this not something that needs to be written into the KIP?
https://github.com/kaiachain/kaia/pull/126/files#diff-12e74de7716895c1d8768e97931b328e64d691bca8f0f27919ffb71623abf671R202-R214

Reply:
I didn't find any execution path where those variables are left uninitialized. If so, that is a bug not complying to the KIP.

Reply:
I agree with you.

The logic will be preserved if you don't forget to set these(storageRoot = EMPTY_ROOT_HASH, codeHash = EMPTY_CODE_HASH, and codeInfo = 0x0) when decorating.

How about specifying the conditions(storageRoot = 0x0, codeHash = nil, and codeInfo = 0x0) when using decoratedFromCompactEncoding?